### PR TITLE
Implement PowerMetrics, PowerFactor, SystemUnbalance Blocks, and Configurable Filters with UncertainValue<T> Support

### PR DESCRIPTION
**Pull request by RalphSteinhagen**
## Summary

This pull request introduces new blocks and features to enhance power system analysis capabilities, along with supporting uncertainty propagation through `UncertainValue<T>`. It also includes minor fixes and improvements across the codebase.

### 1. **PowerMetrics<T, nPhases>**

- **Functionality**: Computes per-phase:
  - Active Power (P)
  - Reactive Power (Q)
  - Apparent Power (S)
  - RMS Voltage (U_rms)
  - RMS Current (I_rms)
- **Features**:
  - Supports both single-phase and multi-phase electrical systems (`nPhases` as a non-type template parameter).
  - Processes voltage and current inputs, applies low-pass filters to calculate average values.
  - Outputs decimated results at a specified rate.

### 2. **PowerFactor<T, nPhases>**

- **Functionality**: Calculates power factor and phase angle for each phase using active power (P) and apparent power (S) inputs.
- **Calculations**:
  - Power Factor: \( \cos(\phi) = \frac{P}{S} \)
  - Phase Angle: \( \phi = \arccos\left(\frac{P}{S}\right) \)
- **Features**:
  - Ensures values are within valid ranges to prevent computational errors.

### 3. **SystemUnbalance<T, nPhases>**

- **Functionality**: Computes:
  - Total Active Power
  - Voltage Unbalance (%)
  - Current Unbalance (%)
- **Calculations**:
  - Unbalance percentages are based on the maximum deviation from average RMS values across all phases.
- **Features**:
  - Suitable for multi-phase systems (requires at least two phases).

### 4. Configurable **BasicFilter<T>**

- **Functionality**: Provides a configurable filtering tool with support for:
  - Filter Types: IIR or FIR
  - Filter Responses: Low-pass, Band-pass, High-pass, Band-stop
  - Modes: Decimating and Non-decimating
- **Features**:
  - Based on user-configurable parameters for flexibility.
  - Supports both fundamental floating-point types and `UncertainValue<T>` types.
- **Benefit**: Enhances signal processing capabilities with uncertainty propagation through the filtering process.

## Support for UncertainValue<T>

- **Integration**: All new blocks (`PowerMetrics`, `PowerFactor`, `SystemUnbalance`, and `BasicFilter`) support both fundamental floating-point types and types wrapped with `UncertainValue<T>`.

## Minor Fixes and Improvements

- **Execution Policy**: Changed default execution policy from `std::execution::seq` to `std::execution::unseq` to potentially improve performance through vectorization.
- **Filter Tools**:
  - Added missing default constructors for `Filter` and `ErrorPropagatingFilter`.
  - Fixed erroneous use of `std::forward<...>` in `ErrorPropagatingFilter`.
- **UncertainValue<T> Enhancements**:
  - Added missing non-uncertainty value returns.
  - Extended support for `sqrt()`, `sin()`, and `cos()` functions to work with all underlying types.
 - **Global Block Registry**: Registered all new blocks, making them available for integration into the system and for use in applications.

## References
- **[0]** IEEE Std 1459-2010: "IEEE Standard Definitions for the Measurement of Electric Power Quantities Under Sinusoidal, Nonsinusoidal, Balanced, or Unbalanced Conditions"
- **[1]** IEC 61000-4-30: "Electromagnetic compatibility (EMC) - Part 4-30: Testing and measurement techniques - Power quality measurement methods"

### CODE REVIEW

File: blocks/electrical/include/gnuradio-4.0/electrical/PowerEstimators.hpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+struct PowerFactor : gr::Block<PowerFactor<T, nPhases>> {
+    using Description = Doc<R""(@brief PowerFactor
+
+Calculates the power factor and phase angle for each phase using active power (P) and apparent power (S) inputs.
+It computes the power factor as cos(𝜙)=P/S, and the phase angle 𝜙=arccos(P/S), ensuring values are within valid ranges.
+)"">;
+    // ports
+    std::vector<PortIn<T>>  P{nPhases}; // active power
+    std::vector<PortIn<T>>  S{nPhases}; // apparent power
+    std::vector<PortOut<T>> powerFactor{nPhases};
```
Comment (drslebedev) [inline blocks/electrical/include/gnuradio-4.0/electrical/PowerEstimators.hpp:134]: Just to keep the naming convention, consider renaming to `power_factor`

File: blocks/filter/include/gnuradio-4.0/filter/time_domain_filter.hpp [hunk 2]
```diff
@@ -107,9 +116,151 @@ a are the feedback coefficients
...
+    PortOut<T> out;
+
+    Annotated<gr::Size_t, "decimation factor", Doc<"Factor by which to downsample/drop input data">, Visible> decim{1};
+
+    GR_MAKE_REFLECTABLE(Decimator, in, out, decim);
+
+    void settingsChanged(const property_map& /*oldSettings*/, const property_map& /*newSettings*/) { this->input_chunk_size = decim; }
+
+    [[nodiscard]] work::Status processBulk(std::span<const T> input, std::span<T> output) noexcept {
+        assert(output.size() >= input.size() / decim);
```
Comment (drslebedev) [inline blocks/filter/include/gnuradio-4.0/filter/time_domain_filter.hpp:244]: I believe that it as guaranteed in Block implementation: `n_inputs = k * input_chunk_size`  and `n_outputs = k * output_chunk_size`

**Review by drslebedev — APPROVED** [review]
Thank you for your PR and the implementation of blocks for power estimation and the updates of filter.

I’m glad to see that we’re moving in a direction by increasingly utilizing standard algorithms and ranges!

The implementation looks solid and well-done, with unit tests effectively covering potential use cases.

I’ve added a few comments suggesting potential improvements to the code—please feel free to review them.

Other than that, the PR is ready to be merged.

**Review by wirew0rm — APPROVED** [review]
Looks very good, looking forward to using these blocks with real-world data.

Added some small issues that I stumbled over.

Formatting should be fixed before merging, but I assume you'll do than anyway on the review pass.

File: blocks/filter/include/gnuradio-4.0/filter/time_domain_filter.hpp [hunk 3]
```diff
@@ -107,9 +116,151 @@ a are the feedback coefficients
...
+
+    // Public settings
+    Annotated<std::string, "filter_type", Doc<"Filter type ('FIR' or 'IIR')">, Visible>                                                filter_type     = std::string(magic_enum::enum_name(_filter_type));
+    Annotated<std::string, "filter_response", Doc<"Filter response ('LOWPASS', 'HIGHPASS', 'BANDPASS', 'BANDSTOP')">, Visible>         filter_response = std::string(magic_enum::enum_name(_filter_response));
+    Annotated<gr::Size_t, "filter_order", Doc<"Filter order">>                                                                         filter_order{3};
+    Annotated<double, "f_low", Doc<"Low cutoff frequency in Hz">, Visible>                                                             f_low{0.1};
+    Annotated<double, "f_high", Doc<"High cutoff frequency in Hz (only for BANDPASS/BANDSTOP)">, Visible>                              f_high{0.2};
+    Annotated<double, "sample rate", Doc<"Sample rate in Hz">, Visible>                                                                sample_rate{1.0};
+    Annotated<gr::Size_t, "decimation factor", Doc<"0: none">>                                                                         decimate{1};
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply blocks/filter/include/gnuradio-4.0/filter/time_domain_filter.hpp:152]: done.

File: blocks/filter/test/qa_FrequencyEstimator.cpp [hunk 4]
```diff
@@ -93,7 +94,7 @@ const boost::ut::suite<"FrequencyEstimatorTests"> FrequencyEstimatorTests = [] {
         testFrequencyEstimator(estimator, processFunc, std::vector<float>(testFrequencies.begin(), testFrequencies.end()), sample_rate, numSamples, noiseAmp, tolerance);
     };
 
-    "Frequency Estimator - Time Domain Decimating"_test = [] {
+    skip / "Frequency Estimator - Time Domain Decimating"_test = [] {
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply blocks/filter/test/qa_FrequencyEstimator.cpp:97]: was disabled by mistake.

**Comment by RalphSteinhagen** [comment]
@drslebedev and @wirew0rm thanks for the review. Adjusted/corrected as much as possible. 

The formatting and skipped tests seem to be unit-test related and would need to be addressed in a follow-up PR. 

The local restyle (clang-format18) produces different results than the CI.

**Comment by RalphSteinhagen** [comment]
@wirew0rm it seems that the new "Restyled" integration does not do a clang-format but additional formatting of the CMakeList.txt files and also files that have been already touched by clang-format. We should update the `formatFiles.sh` and `formatLastCommit.sh` accordingly and maybe add hooks that this is automated.

**Comment by wirew0rm** [comment]
> @wirew0rm it seems that the new "Restyled" integration does not do a clang-format but additional formatting of the CMakeList.txt files and also files that have been already touched by clang-format. We should update the `formatFiles.sh` and `formatLastCommit.sh` accordingly and maybe add hooks that this is automated.

This shouldn't have anything to do with the new integration, it's still running the same restyler with the same configuration, it's just now running inside a github-actions process. It has also always run a handful of other formatters (listed in [.restyled.yaml](https://github.com/fair-acc/gnuradio4/blob/main/.restyled.yaml#L23-L31)). @dantti alredy added pre-commit hooks for some of them [here](https://github.com/fair-acc/gnuradio4/blob/main/.pre-commit-config.yaml).
The two shell scripts should probably be replaced by https://github.com/llvm/llvm-project/blob/main/clang/tools/clang-format/git-clang-format which does the same thing properly and is maintained upstream.

In terms of adding hooks, this will never be fully automatic, since that would be a RCE in git. On the client side we can only make it easy to reformat, ideally with as much standard tools and as few custom scripts as possible. On the CI side now we only provide status and a `git patch` snippet for now, we could also push a separate branch with fixed formatting and/or allow to trigger automatic fixing with a comment `/restyle` command.

**Pull request merged**